include/openthread/tcp.h

```diff
@@ -0,0 +1,589 @@
+ * @param[in]  aSockName  The IP address and port of the host to which to connect.
+ * @param[in]  aFlags     Flags specifying options for this operation (see enumeration above).
+ *
+ * @retval OT_ERROR_NONE    Successfully completed the operation.
+ * @retval OT_ERROR_FAILED  Failed to complete the operation.
+ */
+otError otTcpActiveConnect(otInstance *aInstance, otTcpActiveSocket *aSocket, const otSockAddr *aSockName, uint32_t aFlags);
+
+/**
+ * This enumeration defines flags passed to @p otTcpActiveSendByReference.
+ */
+enum {
+    OT_TCP_ACTIVE_SEND_MORE_TO_COME = 1 << 0,
+};
+
+/**
+ * Adds data referenced by the linked buffer pointed to by @p aBuffer to the
+ * the send buffer. Upon a sucessful call to this function, the linked buffer
+ * and data it references are owned by the TCP stack; they should not be
+ * modified by the application until a "send done" callback returns ownership
+ * of those objects to the application. It is acceptable to call this function
+ * to add another linked buffer to the send queue, even if the "send done"
+ * callback for a previous invocation of this function has not yet fired.
+ *
+ * Note that @p aBuffer should not be chained; its @p mNext field should be
+ * @p NULL. If additional data will be added right after this call, then the
+ * @p OT_TCP_ACTIVE_SEND_MORE_TO_COME flag should be used as a hint to the TCP
+ * implementation.
+ *
+ * @param[in]  aInstance  A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]  aSocket    A pointer to the active TCP socket structure representing the TCP endpoint on which to send data.
+ * @param[in]  aBuffer    A pointer to the linked buffer chain referencing data to add to the send buffer.
+ * @param[in]  aFlags     Flags specifying options for this operation (see enumeration above).
+ *
+ * @retval OT_ERROR_NONE    Successfully added data to the send buffer.
+ * @retval OT_ERROR_FAILED  Failed to add data to the send buffer.
+ */
+otError otTcpActiveSendByReference(otInstance *aInstance, otTcpActiveSocket *aSocket, otLinkedBuffer *aBuffer, uint32_t aFlags);
+
+/**
+ * Adds data to the send buffer by extending the length of the final
+ * otLinkedBuffer in the send buffer by the specified amount. If the send
+ * buffer is empty, then the operation fails.
+ *
+ * @param[in]  aInstance  A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]  aSocket    A pointer to the active TCP socket structure representing the TCP endpoint on which to send data.
+ * @param[in]  aNumBytes  The number of bytes by which to extend the length of the final linked buffer.
+ * @param[in]  aFlags     Flags specifying options for this operation (see enumeration above).
+ *
+ * @retval OT_ERROR_NONE    Successfully added data to the send buffer.
+ * @retval OT_ERROR_FAILED  Failed to add data to the send buffer.
+ */
+otError otTcpActiveSendByExtension(otInstance *aInstance, otTcpActiveSocket *aSocket, size_t aNumBytes, uint32_t aFlags);
```

Comment:
nit: This API is arguably not necessary?

ALP should already know which `otLinkedBuffer` is the last one, otherwise it's unsafe for ALP to call this API. 
Knowing which `otLinkedBuffer` is the last one, ALP can just increase the `mLength`. I think TCP can handle increased `mLength` properly without special handling.

Reply:
I disagree that TCP can handle `mLength` being increased without special handling. Depending on the amount by which `mLength` is increased, TCP may need to send out additional data immediately. As a concrete example, some data in the send buffer may not be sent out immediately due to "silly window avoidance" but a call to `otTcpActiveSendByExtension` causes the threshold to be reached and data to be sent out.